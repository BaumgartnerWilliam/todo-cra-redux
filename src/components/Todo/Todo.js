import React, { memo } from 'react';
import cx from 'classnames';

const ENTER_KEY_CODE = 13;
const onKeyPressBind = fn => event => {
  if (event.which === ENTER_KEY_CODE || event.keyCode === ENTER_KEY_CODE) {
    fn();
    return false;
  }
  return true;
};

// todo: svgs can be extracted to their own component
const Todo = ({ id, done, text, onComplete, onDelete }) => {
  return (
    <li
      aria-label="todo"
      tabIndex="0"
      className={cx(
        'noselect list-group-item list-group-item-action d-flex align-items-center justify-content-between',
        {
          'list-group-item-warning': done
        }
      )}
      onClick={onComplete}
      onKeyPress={onKeyPressBind(onComplete)}>
      {!done && (
        <span aria-label="unchecked" style={{ paddingRight: '16px' }}></span>
      )}
      {done && (
        <span aria-label="checked">
          <svg
            width="1em"
            height="1em"
            viewBox="0 0 16 16"
            className="bi bi-check2"
            fill="currentColor"
            xmlns="http://www.w3.org/2000/svg">
            <path
              fillRule="evenodd"
              d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"
            />
          </svg>
        </span>
      )}
      <span>{text}</span>
      <span
        aria-label="delete-todo"
        tabIndex="0"
        onClick={onDelete}
        onKeyPress={onKeyPressBind(onDelete)}>
        <svg
          width="1em"
          height="1em"
          viewBox="0 0 16 16"
          className="bi bi-trash-fill"
          fill="currentColor"
          xmlns="http://www.w3.org/2000/svg">
          <path
            fillRule="evenodd"
            d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5a.5.5 0 0 0-1 0v7a.5.5 0 0 0 1 0v-7z"
          />
        </svg>
      </span>
    </li>
  );
};

export default memo(Todo);
