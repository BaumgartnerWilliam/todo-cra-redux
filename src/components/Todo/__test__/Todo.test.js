import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Todo from '../Todo';

describe('todo component', () => {
  it('renders without crashing', () => {
    const { container } = render(<Todo />);
    expect(container).toBeInTheDocument();
  });

  it('fires the onClick event properly', () => {
    const todo = {
      id: 12,
      text: 'todo',
      done: false,
      onComplete: jest.fn(),
      onDelete: jest.fn(),
    };
    const { getByLabelText } = render(<Todo {...todo} />);
    getByLabelText('unchecked')
    expect(todo.onComplete).not.toHaveBeenCalled();
    fireEvent.click(getByLabelText('todo'));
    expect(todo.onComplete).toHaveBeenCalled();
  });

  it('fires the onDelete event properly', () => {
    const todo = {
      id: 12,
      text: 'todo',
      done: false,
      onComplete: jest.fn(),
      onDelete: jest.fn(),
    };
    const { getByLabelText } = render(<Todo {...todo} />);
    expect(todo.onDelete).not.toHaveBeenCalled();
    fireEvent.click(getByLabelText('delete-todo'));
    expect(todo.onDelete).toHaveBeenCalled();
  });

  it('matches unchecked snapshot', () => {
    const todo = {
      id: 99,
      text: 'my todo text',
      done: false
    };
    const { container, getByText, getByLabelText } = render(<Todo {...todo} />);

    getByText(todo.text);
    getByLabelText('unchecked')
    expect(container).toMatchSnapshot();
  });

  it('matches checked snapshot', () => {
    const todo = {
      id: 99,
      text: 'my todo text',
      done: true
    };
    const { container, getByText, getByLabelText } = render(<Todo {...todo} />);

    getByText(todo.text);
    getByLabelText('checked')
    expect(container).toMatchSnapshot();
  });
});
