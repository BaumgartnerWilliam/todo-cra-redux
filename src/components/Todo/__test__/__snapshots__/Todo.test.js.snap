// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo component matches checked snapshot 1`] = `
<div>
  <li
    aria-label="todo"
    class="noselect list-group-item list-group-item-action d-flex align-items-center justify-content-between list-group-item-warning"
    tabindex="0"
  >
    <span
      aria-label="checked"
    >
      <svg
        class="bi bi-check2"
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span>
      my todo text
    </span>
    <span
      aria-label="delete-todo"
      tabindex="0"
    >
      <svg
        class="bi bi-trash-fill"
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5a.5.5 0 0 0-1 0v7a.5.5 0 0 0 1 0v-7z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </li>
</div>
`;

exports[`todo component matches unchecked snapshot 1`] = `
<div>
  <li
    aria-label="todo"
    class="noselect list-group-item list-group-item-action d-flex align-items-center justify-content-between"
    tabindex="0"
  >
    <span
      aria-label="unchecked"
      style="padding-right: 16px;"
    />
    <span>
      my todo text
    </span>
    <span
      aria-label="delete-todo"
      tabindex="0"
    >
      <svg
        class="bi bi-trash-fill"
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.5 1a1 1 0 0 0-1 1v1a1 1 0 0 0 1 1H3v9a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2V4h.5a1 1 0 0 0 1-1V2a1 1 0 0 0-1-1H10a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1H2.5zm3 4a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7a.5.5 0 0 1 .5-.5zM8 5a.5.5 0 0 1 .5.5v7a.5.5 0 0 1-1 0v-7A.5.5 0 0 1 8 5zm3 .5a.5.5 0 0 0-1 0v7a.5.5 0 0 0 1 0v-7z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </li>
</div>
`;
